src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModelFactory.java

```diff
@@ -29,19 +37,20 @@ public IntelliJModelFactory(@NotNull Project project) {
 
   @Override
   public Course createCourse(@NotNull String name,
-                             @NotNull List<Module> modules,
-                             @NotNull List<Library> libraries,
-                             @NotNull Map<String, String> requiredPlugins,
-                             @NotNull Map<String, URL> resourceUrls) {
+      @NotNull List<Module> modules,
+      @NotNull List<Library> libraries,
+      @NotNull Map<String, String> requiredPlugins,
+      @NotNull Map<String, URL> resourceUrls) {
     IntelliJCourse course =
         new IntelliJCourse(name, modules, libraries, requiredPlugins, resourceUrls, project);
     // Add a module change listener with the created course instance to the project
     // TODO: These should be handled in those classes; also, renaming issue?
     project.getMessageBus().connect().subscribe(ProjectTopics.MODULES, new ModuleListener() {
       @Override
       public void moduleRemoved(@NotNull Project project,
-                                @NotNull com.intellij.openapi.module.Module projectModule) {
-          course.onComponentRemove(course.getComponentIfExists(projectModule.getName()));
+          @NotNull com.intellij.openapi.module.Module projectModule) {
+        markDependentModulesInvalid(course, projectModule.getName());
```

Comment:
Alternative approach would be that all modules are checked whenever there is a change in module configuration and also on start up. A bit more idiot proof, I'd say. But no need to change now. And how you implemented this is also a more elegant way, so it's fine.

Reply:
I like the idea but suggest to postpone it a bit